CustomKernelInstall: skip if kernel is already installed

Testcase CAPTURE-VHD-BEFORE-TEST fails in installation of custom kernel, if the kernel passed as parameter(-CustomKernel) is already installed  in the VM.
This need to be modified to skip if kernel is already installed and CAPTURE-VHD-BEFORE-TEST should be Passed.

07/17/2019 10:48:53 : [ERROR] Kernel upgrade failed in 1 VMs.
07/17/2019 10:48:53 : [ERROR] Custom Kernel: https://lisbuildstorage04.blob.core.windows.net/kernel/3.10.0-957.21.3.el7/kernel-3.10.0-957.21.3.el7.x86_64.rpm installation FAIL. Aborting tests.

build-CustomKernel Logs:
Web link detected. Downloading https://lisbuildstorage04.blob.core.windows.net/kernel/3.10.0-957.21.3.el7/kernel-3.10.0-957.21.3.el7.x86_64.rpm
Custom Kernel:kernel-3.10.0-957.21.3.el7.x86_64.rpm
Removing packages that do not allow the kernel to be installed
Installing kernel-3.10.0-957.21.3.el7.x86_64.rpm
warning: kernel-3.10.0-957.21.3.el7.x86_64.rpm: Header V3 RSA/SHA256 Signature, key ID fd431d51: NOKEY
Preparing...                          ########################################
	package kernel-3.10.0-957.21.3.el7.x86_64 is already installed

--- a/Libraries/CommonFunctions.psm1
+++ b/Libraries/CommonFunctions.psm1
@@ -574,7 +574,9 @@ function Install-CustomKernel ($CustomKernel, $allVMData, [switch]$RestartAfterU
 				#endregion
 			}
 			$packageInstallJobsRunning = $true
+			$kernelInstalledSkipped = $false
 			$kernelMatchSuccess = "CUSTOM_KERNEL_SUCCESS"
+			$customKernelAlreadyInstall="CUSTOM_KERNEL_ALREADY_INSTALLED"
 			while ($packageInstallJobsRunning) {
 				$packageInstallJobsRunning = $false
 				foreach ( $job in $packageInstallJobs ) {
@@ -589,10 +591,15 @@ function Install-CustomKernel ($CustomKernel, $allVMData, [switch]$RestartAfterU
 						if ( !(Test-Path -Path "$LogDir\$($job.RoleName)-build-CustomKernel.txt" ) ) {
 							Copy-RemoteFiles -download -downloadFrom $job.PublicIP -port $job.SSHPort -files "build-CustomKernel.txt" `
 								-username $user -password $password -downloadTo $LogDir
-							if ( ( Get-Content "$LogDir\build-CustomKernel.txt" ) -imatch $kernelMatchSuccess ) {
+							Rename-Item -Path "$LogDir\build-CustomKernel.txt" -NewName "$($job.RoleName)-build-CustomKernel.txt" -Force | Out-Null
+							if ( ( Get-Content "$LogDir\$($job.RoleName)-build-CustomKernel.txt" ) -imatch $kernelMatchSuccess ) {
 								$kernelSuccess += 1
 							}
-							Rename-Item -Path "$LogDir\build-CustomKernel.txt" -NewName "$($job.RoleName)-build-CustomKernel.txt" -Force | Out-Null
+							elseif ( ( Get-Content "$LogDir\$($job.RoleName)-build-CustomKernel.txt" ) -imatch $customKernelAlreadyInstall ) {
+								Write-LogInfo "Kernel upgrade is skipped in VM."
+								$kernelInstalledSkipped = $true
+								$kernelSuccess +=1
+							}
 						}
 					}
 				}
@@ -601,8 +608,8 @@ function Install-CustomKernel ($CustomKernel, $allVMData, [switch]$RestartAfterU
 				}
 			}
 			if ( $kernelSuccess -eq $jobCount ) {
-				Write-LogInfo "Kernel upgraded to `"$CustomKernel`" successfully in $jobCount VM(s)."
-				if ( $RestartAfterUpgrade ) {
+				if ( $RestartAfterUpgrade -and !$kernelInstalledSkipped ) {
+					Write-LogInfo "Kernel upgraded to `"$CustomKernel`" successfully in $jobCount VM(s)."
 					Write-LogInfo "Now restarting VMs..."
 					if ( $TestProvider.RestartAllDeployments($allVMData) ) {
 						$retryAttempts = 5

--- a/Testscripts/Linux/customKernelInstall.sh
+++ b/Testscripts/Linux/customKernelInstall.sh
@@ -378,6 +378,12 @@ function InstallKernel() {
                 ;;
         esac
         customKernelFilesUnExpanded="${CustomKernel#$LOCAL_FILE_PREFIX}"
+        custom_kernel_version=$(rpm -qp --queryformat='%{Version}-%{Release}.%{Arch}\n' ${customKernelFilesUnExpanded})
+        if [ "$custom_kernel_version" = "$(uname -r)" ]; then
+            LogMsg "CUSTOM_KERNEL_ALREADY_INSTALLED"
+            SetTestStateCompleted
+            exit 0
+        fi
         LogMsg "Removing packages that do not allow the kernel to be installed"
         if [[ "${customKernelFilesUnExpanded}" == *'*.rpm'* ]]; then
             LogMsg "Removing: ${KERNEL_CONFLICTING_PACKAGES}"

